File: 24. Swap Nodes in Pairs.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* swapPairs(struct ListNode* head) {
    struct ListNode dummy, *prev, *first, *second;

    dummy.next = head;
    prev = &dummy;
    first = head;

    while (first && (second = first->next)) {
        first->next = second->next;
        prev->next = second;
        second->next = first;

        prev = first;
        first = first->next;
    }

    return dummy.next;
}

File: Python3/24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummy, dummy.next = ListNode(0), head
        prev, cur = dummy, head

        while cur and cur.next:
            tmp = cur.next.next
            cur.next.next, cur.next, prev.next = cur, tmp, cur.next
            prev, cur = prev.next.next, tmp
        return dummy.next

